First blog about test classloading changes

holly-cummins opened the pull request:
I've added a blog discussing the externals of the test classloading change. 

I was also hoping to discuss the internals, but  that's harder so I kept putting it off. We ought to publish something alongside 3.22, so I've run out of runway to write an epic blog. The discussion of internals should maybe be in its own blog, which gives more time to get "part ii - the terrifying internals" written and out.

maxandersen commented on _posts/2025-04-22-test-classloading-rewrite.adoc:
```diff
@@ -0,0 +1,60 @@
+== What's changing?
+
+The internals of Quarkus test classloading have been rewritten in 3.22.
+In particular, how `@QuarkusTest` test classes are loaded is changing.
```
```suggestion
In particular, this is not affecting Quarkus production classloaders it is only changing how `@QuarkusTest` test classes are loaded during tests.
```

assuming that is a true statement its worth mentioning this is only affecting tests, not prod.

...how about devmode? is it affected ?

holly-cummins replied:
Dev mode isn't changed either. It is only `@QuarkusTest` (not even other test types). I'll add a clarifying sentence to list the exclusions explicitly.

maxandersen replied:
ok, so that devservices start in a different way is *only* for tests not devmode?

holly-cummins replied:
Technically, they start in the same way as they've always done (in the augmentation phase), they just all do it at the same time. :) 

But flippancy aside, I'm not aware of any interactions between these changes and dev mode. "Dev mode" is a bit overloaded because we use it to mean both `quarkus:dev`, which actually launches two applications, and also the "not-test" application part of that. So in `quarkus:dev` the tests will get run in a different way, but the application side should be the same. I wondered if the frontloaded dev services could conflict with the application-side-dev-service, but I don't think that's an issue, because if there was a possibility for conflict on port numbers, it would have already manifested halfway through a test run, in the old architecture.

cescoffier replied:
No, the change only affects test, not dev. However, it affects continuous testing happening in dev mode.

cescoffier replied:
I would write it using multiple sentences:

The internals of Quarkus test classloading have been rewritten in 3.22.
It does **not** affect production and dev modes. However, it changes how `@QuarkusTest`, `@QuarkusIntegrationTest`, and other Quarkus tests are loaded.

holly-cummins replied:
I've gone with the last one, as suggested.

maxandersen commented on _posts/2025-04-22-test-classloading-rewrite.adoc:
```diff
@@ -0,0 +1,60 @@
+
+The internals of Quarkus test classloading have been rewritten in 3.22.
+In particular, how `@QuarkusTest` test classes are loaded is changing.
+
```
would be good to give a slight more context up front stating that we know this could cause some tests that relied on sideeffects of the old way to fail - we expect most due to actual bugs just not discoverable before and are looking for feedback and below blog outlines the details.

holly-cummins replied:
Done!

maxandersen commented on _posts/2025-04-22-test-classloading-rewrite.adoc:
```diff
@@ -0,0 +1,60 @@
+One of the goals of this change was that the rewrite didn't touch any tests in our test suite, to make sure they'd all continue working without updates.
+In practice, there have been a few hiccups and we've discovered some edge cases in the broader ecosystem.
+
+=== Dropped support
```
the lines before here talks about the known regressions - should this dropped support section move above or below that so there is a natural flow ?

maxandersen commented on _posts/2025-04-22-test-classloading-rewrite.adoc:
```diff
@@ -0,0 +1,60 @@
+
+=== Known regressions
+
+- *Dev services start in the discovery phase*. Quarkus Dev Services are currently started during the augmentation phase. With the new design, all augmentation happens at the beginning of the test run. If several test classes with different Dev Service configuration are augmented before any tests are run, multiple differently-configured Dev Services may be running at the same time. This can cause  ...
```
are we sure we can keep that promise of fixing that? my understanding from the conversation was that this requirement of starting during augmentation phase would be too harsh for devservices ?

maxandersen commented on _posts/2025-04-22-test-classloading-rewrite.adoc:
```diff
@@ -0,0 +1,60 @@
+
+- *Dev services start in the discovery phase*. Quarkus Dev Services are currently started during the augmentation phase. With the new design, all augmentation happens at the beginning of the test run. If several test classes with different Dev Service configuration are augmented before any tests are run, multiple differently-configured Dev Services may be running at the same time. This can cause  ...
+- *Config access from JUnit conditions*. Using a `ConfigProvider` from a custom JUnit condition will https://github.com/quarkusio/quarkus/issues/47081[trigger a `ServiceConfigurationError`]. The workaround is to set the thread context classloader to `this.getClass().getClassLoader()` before reading config, and then set it back afterwards.
+- Increased memory footprint running tests. For suites using multiple profiles and resources, more heap or metaspace may be needed.
```
why? I would have expected less usage as there is just one classloader ...or is it because before there was one shared "heavy" classloader and many junit classloaders and now there are just many "heavy" junit classloaders ?

holly-cummins replied:
The overall number of classloaders is the same, and they're basically the same size as before. The extra pressure on the heap happens because we have more of them present on the heap at the same time. It used to be we'd close one before creating the next. It's a side effect of frontloading the classloading.

maxandersen replied:
gotcha....but they will all get closed/collected per test class or suite? i assume they must be cleaned up as otherwise continues testing will have a hard time :)

holly-cummins replied:
(We collect them as soon as we're done with the test/test profile group, same as before.)

maxandersen commented on _posts/2025-04-22-test-classloading-rewrite.adoc:
```diff
@@ -0,0 +1,60 @@
+
+=== Things to watch out for
+
+- *Test order change*. As part of the rewrite, the execution order of some tests has swapped around. Of course, we all know tests should not depend on execution order if they don't set an order explicitly. However, it's easy to not notice that a test requires a certain order... until the order changes. We discovered some tests in our own suite that were sensitive to the execution order, and other ...
```
just curious - what order was it before and what is it now ?

maxandersen replied:
okey, so the order is the same between new runs ...just not same as old runs. cool.

holly-cummins replied:
That's better, thanks!

maxandersen reviewed (changes requested):
really good write up - added a few clarifying suggestions but good stuff!

holly-cummins commented on _posts/2025-04-22-test-classloading-rewrite.adoc:
```diff
@@ -0,0 +1,60 @@
+=== Known regressions
+
+- *Dev services start in the discovery phase*. Quarkus Dev Services are currently started during the augmentation phase. With the new design, all augmentation happens at the beginning of the test run. If several test classes with different Dev Service configuration are augmented before any tests are run, multiple differently-configured Dev Services may be running at the same time. This can cause  ...
+- *Config access from JUnit conditions*. Using a `ConfigProvider` from a custom JUnit condition will https://github.com/quarkusio/quarkus/issues/47081[trigger a `ServiceConfigurationError`]. The workaround is to set the thread context classloader to `this.getClass().getClassLoader()` before reading config, and then set it back afterwards.
```
I have a simple, elegant, clearly correct fix for this. Maddeningly, the fix breaks gradle, which does unorthodox things in its test classloading. I haven't yet got a fix that works for both gradle and non-gradle paths, but again, fingers crossed for next release.

If people do the TCCL workaround it will be a no-op once my fix drops. So better to remove it, because of code clutter, but harmless.

cescoffier commented on _posts/2025-04-22-test-classloading-rewrite.adoc:
```diff
@@ -0,0 +1,63 @@
+
+The internals of Quarkus test classloading have been rewritten in 3.22.
+In particular, how `@QuarkusTest` test classes are loaded is changing.
+Other test extensions haven't changed, nor has dev mode and production mode classloading.
```
If we say it before, we can remove this sentence.

holly-cummins replied:
@maxandersen suggested the sentence, to make it very clear to people who might be skimming that production mode wasn't changing.

cescoffier replied:
I will go with the last one.

cescoffier commented on _posts/2025-04-22-test-classloading-rewrite.adoc:
```diff
@@ -0,0 +1,63 @@
+The internals of Quarkus test classloading have been rewritten in 3.22.
+In particular, how `@QuarkusTest` test classes are loaded is changing.
+Other test extensions haven't changed, nor has dev mode and production mode classloading.
+This change should make Quarkus testing work better, and it allowed us to fix a pile of longstanding bugs.
```
...and improve the integration with test frameworks such as Pact.

cescoffier commented on _posts/2025-04-22-test-classloading-rewrite.adoc:
```diff
@@ -0,0 +1,63 @@
+In previous versions, Quarkus tests were invoked using the default JUnit classloader, and then executed in a different, Quarkus-aware, classloader.
+
+This mostly worked very well, and meant that `QuarkusTest` tests mostly behaved as if they were part of the same application as the code under test.
+The Quarkus test framework could start and stop Quarkus instances at the right point in the test lifecycle, inject CDI dependencies, and do other useful Quarkus bytecode manipulation.
```
point or time?

holly-cummins replied:
I'd say point, because we don't judge it by a clock, we do it relative to other events in the lifecycle. But I don't have strong feelings, if you think time is clearer?

cescoffier replied:
point is fine - I was just wondering.

cescoffier commented on _posts/2025-04-22-test-classloading-rewrite.adoc:
```diff
@@ -0,0 +1,63 @@
+
+This mostly worked very well, and meant that `QuarkusTest` tests mostly behaved as if they were part of the same application as the code under test.
+The Quarkus test framework could start and stop Quarkus instances at the right point in the test lifecycle, inject CDI dependencies, and do other useful Quarkus bytecode manipulation.
+However, some use cases didn't work. Tests using advanced JUnit features like `@TestTemplate` and `@ParameterizedTest` sometimes found that the same test code might appear to run in several classloaders in a single test, or that injected dependencies weren't always available.
```
should we say Junit 5 features ?

cescoffier commented on _posts/2025-04-22-test-classloading-rewrite.adoc:
```diff
@@ -0,0 +1,63 @@
+
+While Quarkus extensions can do all sorts of marvellous bytecode manipulation to improve the developer experience, they cannot manipulate test classes with the same freedom that they do normal application classes.
+
+Over time, test-related defects were building up that couldn't be changed without a fundamental rewrite of test classloading. The Quarkus test code itself was also growing ever-more complex as it tried to work around various JUnit edge cases. Moving test instances from one classloader to another involved serializing and deserialization, which is harder to implement on newer JVM versions with tigh ...
```
test classloading only? Or test execution too?

cescoffier commented on _posts/2025-04-22-test-classloading-rewrite.adoc:
```diff
@@ -0,0 +1,63 @@
+
+From Quarkus 3.22 onwards, this is exactly how `@QuarkusTest` classloading works.
+What do your tests need to change in order to work with the new architecture?
+Hopefully, nothing.
```
I would write it: 

**nothing** (hopefully)

But it reduces the level of confidence.

cescoffier commented on _posts/2025-04-22-test-classloading-rewrite.adoc:
```diff
@@ -0,0 +1,63 @@
+
+=== Known regressions
+
+- *Dev services start in the discovery phase*. Quarkus Dev Services are currently started during the augmentation phase. With the new design, all augmentation happens at the beginning of the test run. If several test classes with different Dev Service configuration are augmented before any tests are run, multiple differently-configured Dev Services may be running at the same time. This can cause  ...
```
discovery / augmentation - we might need a link to a reference guide. I actually never heard about the discovery phase.

cescoffier replied:
We may have to explain augmentation somewhere (in an admonition)

holly-cummins replied:
That's JUnit discovery (oops!). It's a problem with working with JUnit that loads of concepts get overloaded - there's JUnit extensions and Quarkus extensions, JUnit phases and Quarkus phases ... which kind of phase was clear in my head, but no one else would have any way of knowing, so I'll clarify.

I'll also add some links for augmentation.

cescoffier commented on _posts/2025-04-22-test-classloading-rewrite.adoc:
```diff
@@ -0,0 +1,63 @@
+
+- *Test order change*. A side effect of the rewrite is that the execution order of test can be different. Of course, we all know tests should not depend on execution order if they don't set an order explicitly. However, it's easy to not notice that a test requires a certain order... until the order changes. We discovered some tests in our own suite that were sensitive to the execution order, and  ...
+- *Test timing change*. We also discovered that the rewrite exposed some timing issues in tests. Because classloading is frontloaded at the beginning of the test run, rather than between test executions, there's less time for asynchronous operations to finish between tests. For example, there may no longer be time for external state to 'reset' before the next test starts. This might expose some h ...
+
```
Both are hidden bugs in the test suites or Quarkus, so I would say that's a good thing.

holly-cummins replied:
Agreed. We're Helping! 😁

cescoffier commented on _posts/2025-04-22-test-classloading-rewrite.adoc:
```diff
@@ -0,0 +1,63 @@
+=== Dropped support
+
+- *`@TestProfile` on `@Nested` tests.* Mixing different test profiles and test resources on `@Nested` tests is no longer supported. By definition, every `@TestProfile` must get its own Quarkus application and classloader. Having multiple classloaders execute one test isn't compatible with loading the test with the classloader used to run it.
+- *Version 2.x of the Maven Surefire plugin*. Versions below 3.x of the Maven Surefire plugin will no longer work with `@QuarkusTest`. Version 3 of the Surefire plugin was released in 2023, so version 2 is now rather old.
```
Do we document that in the breaking change? I know it's very old and should not be used, but it might be a problem for some users.

cescoffier commented on _posts/2025-04-22-test-classloading-rewrite.adoc:
```diff
@@ -0,0 +1,63 @@
+More excitingly, test-related extensions, like the Pact extensions, can now add new features to reduce test boilerplate.
+
+As always, if you spot issues or oddities, please let us know on https://quarkusio.zulipchat.com/[zulip] or https://github.com/quarkusio/quarkus/issues[raise an issue].
+
```
I would mention the working group somewhere :-)

holly-cummins replied:
Yes, good point! I had every intention of mentioning it, and even pasted the link into a draft, but lost it. Will do.

cescoffier reviewed (changes requested):
Looks great! I added a few comments, nothing serious.

cescoffier commented:
Agreed, the admonition is too prominent. It's where I would love to have "side notes" - link it is.

maxandersen commented:
mark as approved based on the issues i commented on is fixed. let you decide when to merge after admonition fixed.

cescoffier reviewed (approved):
Great! 

I would postpone the publication after the 3.22 platform release announcement, maybe just after it. But, no strong opinion.

holly-cummins commented:
I've adjusted the date and squashed, so we should be good to merge tomorrow. I've had one zulip message reporting the dev services startup sequence change already, so we should not delay past tomorrow, I think.
